RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc

```diff
@@ -0,0 +1,776 @@
+      auto trackIdx = t.seedIndex();
+      auto const &track = tracks[trackIdx];
+      if (!usedSeeds[trackIdx] and t.raw_energy() > 0) {
+        usedSeeds[trackIdx] = true;
+        usedTrackstersMerged[mergedIdx] = true;
+
+        std::vector<int> trackstersTRKwithSameSeed;
+        std::vector<int> trackstersTRKEMwithSameSeed;
+
+        for (const auto &tracksterIterationPair : seedToTracksterAssociator[trackIdx]) {
+          if (tracksterIterationPair.first != mergedIdx and !usedTrackstersMerged[tracksterIterationPair.first] and
+              trackstersMergedHandle->at(tracksterIterationPair.first).raw_energy() > 0.) {
+            if (tracksterIterationPair.second == TracksterIterIndex::TRKEM) {
+              trackstersTRKEMwithSameSeed.push_back(tracksterIterationPair.first);
+            } else if (tracksterIterationPair.second == TracksterIterIndex::TRKHAD) {
+              trackstersTRKwithSameSeed.push_back(tracksterIterationPair.first);
+            }
+          }
+        }
+
+        float tracksterTotalRawPt = t.raw_pt();
+        std::vector<int> haloTrackstersTRKIdx;
+        bool foundCompatibleTRK = false;
+
+        for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+          usedTrackstersMerged[otherTracksterIdx] = true;
+          tracksterTotalRawPt += trackstersMergedHandle->at(otherTracksterIdx).raw_pt();
+
+          // Check the X,Y,Z barycenter and merge if they are very close (halo)
+          if ((t.barycenter() - trackstersMergedHandle->at(otherTracksterIdx).barycenter()).mag2() <
+              halo_max_distance2_) {
+            haloTrackstersTRKIdx.push_back(otherTracksterIdx);
+
+          } else {
+            foundCompatibleTRK = true;
+          }
+        }
+
+        //check if there is 1-to-1 relationship
+        if (trackstersTRKEMwithSameSeed.empty()) {
+          if (foundCompatibleTRK) {
+            TICLCandidate tmpCandidate;
+            tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
+            double raw_energy = t.raw_energy();
+
+            tmpCandidate.setCharge(track.charge());
+            tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+            tmpCandidate.setPdgId(211 * track.charge());
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+              raw_energy += trackstersMergedHandle->at(otherTracksterIdx).raw_energy();
+            }
+            tmpCandidate.setRawEnergy(raw_energy);
+            math::XYZTLorentzVector p4(raw_energy * track.momentum().unit().x(),
+                                       raw_energy * track.momentum().unit().y(),
+                                       raw_energy * track.momentum().unit().z(),
+                                       raw_energy);
+            tmpCandidate.setP4(p4);
+            resultCandidates->push_back(tmpCandidate);
```

Comment:
consider reducing the repetition with respect to the code below by using a function

Reply:
I think in this case a function would make the code readability harder, I will rearrange the code a bit to avoid duplication as much as possible, but I think it would be better to have all the logic to build the candidates more explicit.

Reply:
```C++
if (foundCompatibleTRK) {
  resultCandidates->push_back(buildChargedHadron(...));
}
} else {
  resultCandidates->push_back(buildElectron(...));
}
...
```
I would suggest something like this here and below. Having code blocks that are hundreds of lines long seems difficult to read and maintain later.

Reply:
Hello @jpata, thanks for the comments! I think implementing functions is slightly unnecessary here:
- this is part of the TICL v3 logic which is deprecated and will be removed soon (and indeed this code is already in production as `TrackstersMergeProducer.cc`, we just changed the name of the file to keep it there till everything switches to using TICL v4)
- It would require some non-trivial modification of the code, and we risk introducing bugs in something which is already very well tested